packages/plugin-ext/src/common/rpc-protocol.ts

```diff
@@ -131,416 +117,154 @@ export class RPCProtocolImpl implements RPCProtocol {
-    }
-
-    private invokeHandler(proxyId: string, methodName: string, args: any[]): Promise<any> {
-        try {
-            return Promise.resolve(this.doInvokeHandler(proxyId, methodName, args));
-        } catch (err) {
-            return Promise.reject(err);
-        }
-    }
-
-    private doInvokeHandler(proxyId: string, methodName: string, args: any[]): any {
-        const actor = this.locals.get(proxyId);
-        if (!actor) {
-            throw new Error('Unknown actor ' + proxyId);
-        }
-        const method = actor[methodName];
-        if (typeof method !== 'function') {
-            throw new Error('Unknown method ' + methodName + ' on actor ' + proxyId);
-        }
-        return method.apply(actor, args);
-    }
-
-    private cancel(req: string): string {
-        return `{"type":${MessageType.Cancel},"id":"${req}"}`;
-    }
-
-    private request(req: string, rpcId: string, method: string, args: any[]): string {
-        return `{"type":${MessageType.Request},"id":"${req}","proxyId":"${rpcId}","method":"${method}","args":${JSON.stringify(args, this.replacer)}}`;
-    }
-
-    private replyOK(req: string, res: any): string {
-        if (typeof res === 'undefined') {
-            return `{"type":${MessageType.Reply},"id":"${req}"}`;
-        }
-        return `{"type":${MessageType.Reply},"id":"${req}","res":${safeStringify(res, this.replacer)}}`;
-    }
+            this.locals.set(identifier.id, handler);
+            if (Disposable.is(instance)) {
+                this.toDispose.push(instance);
+            }
+            this.toDispose.push(Disposable.create(() => this.locals.delete(identifier.id)));
 
-    private replyErr(req: string, err: any): string {
-        err = typeof err === 'string' ? new Error(err) : err;
-        if (err instanceof Error) {
-            return `{"type":${MessageType.ReplyErr},"id":"${req}","err":${safeStringify(transformErrorForSerialization(err))}}`;
         }
-        return `{"type":${MessageType.ReplyErr},"id":"${req}","err":null}`;
+        return instance;
     }
 }
 
-function canceled(): Error {
-    const error = new Error('Canceled');
-    error.name = error.message;
-    return error;
-}
-
 /**
  * Sends/Receives multiple messages in one go:
  *  - multiple messages to be sent from one stack get sent in bulk at `process.nextTick`.
  *  - each incoming message is handled in a separate `process.nextTick`.
  */
-class RPCMultiplexer implements Disposable, MessageConnection {
-
-    private readonly connection: MessageConnection;
-    private readonly sendAccumulatedBound: () => void;
-
-    private messagesToSend: string[];
-
-    private readonly messageEmitter = new Emitter<string>();
-    private readonly toDispose = new DisposableCollection();
-
-    constructor(connection: MessageConnection) {
-        this.connection = connection;
-        this.sendAccumulatedBound = this.sendAccumulated.bind(this);
+export class QueuingChannelMultiplexer extends ChannelMultiplexer {
```

Comment:
I really feel we should keep the concerns of "batching messages" and "multiplexing channels" separate. The old `RPCMultiplexer` name was always wrong, since it never did multiplex, it batched.

Reply:
So I would really have a `BatchingChannel` that takes care of batching and use a stock `ChannelMultiplexer` on top of that.

Reply:
Yes you are absolutely right. A batching channel is definitely the cleaner solution.

Reply:
Fixed with https://github.com/eclipse-theia/theia/commit/2b4016e8da50466f271bce2b2f808bf0ed371a10